tensorly/backend/core.py

```diff
@@ -593,17 +593,17 @@ def kr(self, matrices):
                              'given.'.format(len(matrices)))
 
         n_col = self.shape(matrices[0])[1]
-        for i, e in enumerate(matrices[1:]):
-            if not i:
-                res = matrices[0]
-            s1, s2 = self.shape(res)
-            s3, s4 = self.shape(e)
-            if not s2 == s4 == n_col:
+        if any(i.shape[1] != n_col for i in matrices):
                 raise ValueError('All matrices should have the same number of columns.')
-            a = self.reshape(res, (s1, 1, s2))
-            b = self.reshape(e, (1, s3, s4))
-            res = self.reshape(a * b, (-1, n_col))
-        return res
+        shapes = [(1,)*i + (s,) + (1,)*(len(matrices)-i-1) + (n_col,) for i, s in
+                  enumerate([m.shape[0] for m in matrices])]
+        reshaped_matrices = [self.reshape(i, s) for i, s in zip(matrices, shapes)]
```

Comment:
Is there any performance gain? I find it significantly harder to read.

Reply:
I don't remember the details of why I did this. I think it has to do with being able to inject the mask into the calculation.